ipsec/sa.c

```diff
@@ -0,0 +1,512 @@
+			if (mask_len2 > entry_mask_len)
+				continue;
+
+			entry_port = sort_src_first ? entry->src_port : entry->dest_port;
+			if (port != IPSEC_PORT_ANY && entry_port == IPSEC_PORT_ANY)
+				break;
+
+			entry_port = sort_src_first ? entry->dest_port : entry->src_port;
+			if (port2 != IPSEC_PORT_ANY && entry_port == IPSEC_PORT_ANY)
+				break;
+		}
+
+		if (entry == NULL)
+			LIST_ADD_ELEM(table->first, n, list);
+		else {
+			LIST_ADD_ELEM(entry, n, list);
+			if (table->first == entry)
+				table->first = n;
+		}
+	}
+	mutexUnlock(table->mutex);
+	IPSEC_LOG_TRC(IPSEC_TRACE_RETURN, "return = %p", n);
+	return n;
+}
+
+
+void ipsec_spd_del(spd_entry_t *entry, spd_table *table)
+{
+	mutexLock(table->mutex);
+	LIST_REMOVE(table, entry, list);
+	free(entry);
+	mutexUnlock(table->mutex);
+}
+
+
+int ipsec_spd_del_maybe(spd_entry_t *ep, spd_table *table)
+{
+	spd_entry_t *entry;
+	int ret = -ENOENT;
+
+	mutexLock(table->mutex);
+	LIST_FOR_EACH(table, entry, list)
+	{
+		if (entry != ep)
+			continue;
+
+		LIST_REMOVE(table, entry, list);
+		free(entry);
+		ret = 0;
+		break;
+	}
+	mutexUnlock(table->mutex);
+
+	return ret;
+}
+
+
+spd_entry_t *ipsec_spd_lookup(void *payload, spd_table *table, unsigned flags)
+{
+	spd_entry_t *entry;
+	struct ip_hdr *header = (struct ip_hdr *)payload;
+	ipsec_in_ip_t *ip = (ipsec_in_ip_t *)payload;
+	int ignore_src = flags & IPSEC_MATCH_DST;
+	int ignore_dst = flags & IPSEC_MATCH_SRC;
+
+	//	IPSEC_LOG_TRC(IPSEC_TRACE_ENTER, "header=%p, table=%p, [src=0x%08lx, dst=0x%08lx]",
+	//		(void *)ip, (void *)table, header->src.addr, header->dest.addr);
+
+	mutexLock(table->mutex);
+	LIST_FOR_EACH(table, entry, list)
+	{
+		if ((ignore_src || ipsec_ip_addr_maskcmp(header->src.addr, entry->src, entry->src_mask)) &&
+			(ignore_dst || ipsec_ip_addr_maskcmp(header->dest.addr, entry->dest, entry->dest_mask))) {
+
+			if (entry->protocol == IPSEC_PROTO_ANY)
+				break;
+			if (entry->protocol == IPH_PROTO(header)) {
+				if (IPH_PROTO(header) == IP_PROTO_TCP) {
+					if ((entry->src_port == IPSEC_PORT_ANY) || (entry->src_port == ip->inner_header.tcp.src))
+						if ((entry->dest_port == 0) || (entry->dest_port == ip->inner_header.tcp.dest)) {
+							// IPSEC_LOG_TRC(IPSEC_TRACE_RETURN, "tmp_entry = %p", entry );
+							break;
+						}
+				}
+				else if (IPH_PROTO(header) == IP_PROTO_UDP) {
+					if ((entry->src_port == IPSEC_PORT_ANY) || (entry->src_port == ip->inner_header.udp.src))
+						if ((entry->dest_port == 0) || (entry->dest_port == ip->inner_header.udp.dest)) {
+							// IPSEC_LOG_TRC(IPSEC_TRACE_RETURN, "tmp_entry = %p", entry );
+							break;
+						}
+				}
+				else
+					break;
+			}
+		}
+	}
+	mutexUnlock(table->mutex);
+	// IPSEC_LOG_TRC(IPSEC_TRACE_RETURN, "return = %p", entry );
+	return entry;
+}
+
+
+sad_entry_t *ipsec_sad_add(const sad_entry_t *entry, sad_table *table)
+{
+	sad_entry_t *n;
+
+	n = malloc(sizeof(sad_entry_t));
+	if (n == NULL)
+		return NULL;
+
+	memcpy(n, entry, sizeof(sad_entry_t));
```

Comment:
We don't need to lock somehow access to `entry`?

Reply:
entry is local data.